### PR TITLE
Implement reference counting in MPT

### PR DESCRIPTION
**Pull request by fyrchik**
Results for a full testnet dump (5003730 blocks):
```
# Master-2.x
real    106m36,227s
user    192m32,855s
sys     14m45,739s
13G     .

# Refcount
real    121m10,446s
user    231m6,476s
sys     17m0,014s
9,0G    .

# Generation GC (note: GC hadn't been finished when restore was completed)
real    123m35,329s
user    284m21,860s
sys     22m32,090s
9,5G    .
```

First 2M mainnet blocks (will publish 2M-3M later). Note, generation GC starts every 200k blocks, so not all garbage was collected).
```
# Generation GC
real    46m31,497s
user    76m51,146s
sys     5m52,001s
12G     .

# Master-2.x
real    44m28,469s
user    74m43,983s
sys     5m20,414s
13G     .

# Refcount
real    44m22,258s
user    74m16,260s
sys     5m5,966s
11G     .
```

First 3M mainnet blocks (same note for generational GC):
```
# Master-2.x
real    644m50,778s
user    1223m8,365s
sys     69m23,930s
56G     chains

# Generational
real    698m50,998s
user    1019m11,077s
sys     85m33,651s
46G     chains

# Refcount
real    510m3,492s
user    1167m14,155s
sys     40m38,908s
28G     chains
```

This can be used for benchmark, though `VerifyBlocks: false` must be set manually. https://gist.github.com/fyrchik/df3733df056cdf9c0ab6b39dfc6d759d

UPDATE:
Results for optimized refcount version (boltdb was used, leveldb TBD):
```
# Refcount (stored at the end of node, no optimizations)
real    106m48,438s
user    151m50,943s
sys     6m56,261s
16G     chains

# Master
real    117m21,565s
user    145m42,361s
sys     10m26,123s
22G     chains

# Refcount + flush only changed
real    106m49,477s
user    150m45,168s
sys     6m54,792s
16G     chains

# Refcount + flush only changed + cache refcount of nodes read from store
real    107m5,384s
user    151m29,465s
sys     6m59,464s
16G     chains
```

UPDATE:
Results for leveldb+testnet with config switch
```
# Master, no GC
real    106m4,500s
user   193m19,865s
sys     14m41,085s
12G chains

# GC disabled in config
real   108m51,511s
user  199m24,728s
sys    14m56,655s
12G	chains

# GC enabled
real    119m48,806s
user   229m39,195s
sys    16m48,748s
8,3G	chains
```

### CODE REVIEW

**Comment by roman-khimov** [comment]
So, 31% better in size and 13% worse in time for testnet.

**Comment by fyrchik** [comment]
@roman-khimov time difference should be even less if we enable block and tx verification.

File: pkg/core/mpt/node.go [hunk 1]
```diff
@@ -27,6 +27,17 @@ type NodeObject struct {
 	Node
 }
 
+// extendedNodeObject extends NodeObject with gc-related info.
+type extendedNodeObject struct {
+	NodeObject
+	RefCount uint32
+}
+
+type nodeWrapper interface {
```
Comment (roman-khimov) [inline pkg/core/mpt/node.go:36]: I think think interface can be avoided, and better be avoided (as it adds some overhead). You can add `RefCount` to `NodeObject` and deserialize it in `getFromStore` when needed.

File: pkg/core/mpt/trie.go [hunk 2]
```diff
@@ -13,7 +15,10 @@ import (
 type Trie struct {
 	Store *storage.MemCachedStore
 
-	root Node
+	gcEnabled bool
+
```
Comment (roman-khimov) [inline pkg/core/mpt/trie.go:19]: Useless blank line.

File: pkg/core/mpt/trie.go [hunk 3]
```diff
@@ -13,7 +15,10 @@ import (
 type Trie struct {
 	Store *storage.MemCachedStore
 
-	root Node
+	gcEnabled bool
```
Comment (roman-khimov) [inline pkg/core/mpt/trie.go:18]: s/gc/refcount/

File: pkg/core/mpt/trie.go [hunk 4]
```diff
@@ -348,14 +464,23 @@ func (t *Trie) getFromStore(h util.Uint256) (Node, error) {
...
+	}
 	n.DecodeBinary(r)
 	if r.Err != nil {
 		return nil, r.Err
 	}
-	n.Node.(flushedNode).setCache(data, h)
-	return n.Node, nil
+
+	if t.gcEnabled {
+		data = data[:len(data)-4]
```
Comment (roman-khimov) [inline pkg/core/mpt/trie.go:441]: I think if you're to avoid this cut you can then avoid `NewBufBinWriter` in flush() and that would be really helpful.

File: pkg/core/mpt/trie.go [hunk 5]
```diff
@@ -314,32 +354,108 @@ func makeStorageKey(mptKey []byte) []byte {
...
+		no.EncodeBinary(w.BinWriter)
+		data = w.Bytes()
+	}
+	cnt += t.updated[h]
+	switch {
+	case cnt < 0:
+		// BUG: negative reference count
+		// FIXME remove after debugging
+		panic(fmt.Sprintf("negative reference count: %s new %d, upd %d", h.StringBE(), cnt, t.updated[h]))
+	case cnt == 0:
```
Comment (roman-khimov) [inline pkg/core/mpt/trie.go:404]: You can avoid encoding the node in this case.

File: config/protocol.mainnet.yml [hunk 6]
```diff
@@ -30,7 +30,7 @@ ProtocolConfiguration:
     IssueTransaction: 500
     PublishTransaction: 500
     RegisterTransaction: 10000
-  VerifyBlocks: true
+  VerifyBlocks: false
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply config/protocol.mainnet.yml:33]: Fixed.

File: pkg/config/protocol_config.go [hunk 7]
```diff
@@ -22,6 +22,10 @@ type (
 		AddressVersion byte `yaml:"AddressVersion"`
 		// EnableStateRoot specifies if exchange of state roots should be enabled.
 		EnableStateRoot bool `yaml:"EnableStateRoot"`
+		// EnableMPTGC specifies if MPT GC should be enabled.
+		// If true, DB size will be smaller, but older roots won't be accessible.
+		// This value should remain the same for the same database.
+		EnableMPTGC bool `yaml:"EnableMPTGC"`
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/config/protocol_config.go:28]: Fixed.

File: pkg/core/dao/dao.go [hunk 8]
```diff
@@ -510,16 +510,27 @@ func makeStateRootKey(height uint32) []byte {
...
+		dao.MPT = mpt.NewTrie(nil, enableGC, dao.Store)
+		var val byte
+		if enableGC {
+			val = 1
+		}
+		return dao.Store.Put(gcKey, []byte{val})
+	}
+	v, err := dao.Store.Get(gcKey)
+	if err != nil {
+		return fmt.Errorf("MPT GC info not found: %w", err)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/dao/dao.go:525]: Fixed (keep old state by default).

File: pkg/core/mpt/node.go [hunk 9]
```diff
@@ -27,6 +27,17 @@ type NodeObject struct {
 	Node
 }
 
+// extendedNodeObject extends NodeObject with gc-related info.
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/mpt/node.go:30]: Fixed.

File: pkg/core/mpt/node.go (hunk 1 above)
Comment (fyrchik) [reply pkg/core/mpt/node.go:36]: Fixed.

File: pkg/core/mpt/trie.go (hunk 3 above)
Comment (fyrchik) [reply pkg/core/mpt/trie.go:18]: Fixed.

File: pkg/core/mpt/trie.go (hunk 4 above)
Comment (fyrchik) [reply pkg/core/mpt/trie.go:441]: There is no encoding in subsequent commits. `append` in `updateRefCount` is done to this slice with a sufficient capacity.

File: pkg/core/mpt/trie.go [hunk 10]
```diff
@@ -314,32 +359,69 @@ func makeStorageKey(mptKey []byte) []byte {
...
+			if t.refcountEnabled {
+				t.updateRefCount(h, cnt, node)
+			} else if cnt > 0 {
+				_ = t.Store.Put(makeStorageKey(h.BytesBE()), node)
+			}
+		}
+	}
+	t.initial = make(map[util.Uint256]int)
+	t.updated = make(map[util.Uint256]int)
+	t.nodes = make(map[util.Uint256][]byte)
```
Comment (roman-khimov) [inline pkg/core/mpt/trie.go:377]: But `initial` and `nodes` data can still be left until compaction happens.

**Review by roman-khimov — CHANGES_REQUESTED** [review]
1) Commits became a little messy now, 712a4c3ea2ca88dcd52448a7a5cd1dd92a46aaad probably doesn't work without others.
2) We need real tests with real edge cases (duplicate nodes in the tree, creating a new node that already exists, making no-op changes before flush) for ref counting

File: pkg/core/mpt/trie.go [hunk 11]
```diff
@@ -18,6 +18,7 @@ type Trie struct {
 	root            Node
 	refcountEnabled bool
 	nodes           map[util.Uint256][]byte
+	initial         map[util.Uint256]int
```
[the remark this answers is not in the document]
Comment (roman-khimov) [reply pkg/core/mpt/trie.go:21]: And it should help cache locality.

File: pkg/core/dao/dao.go [hunk 12]
```diff
@@ -510,16 +510,28 @@ func makeStateRootKey(height uint32) []byte {
...
+			val = 1
+		}
+		return dao.Store.Put(gcKey, []byte{val})
+	}
+	var hasGC bool
+	if v, err := dao.Store.Get(gcKey); err == nil {
+		hasGC = v[0] != 0
+	}
+	if hasGC != enableRefCount {
+		return fmt.Errorf("MPT GC setting mismatch: old=%v, new=%v", hasGC, enableRefCount)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/dao/dao.go:528]: Fixed.

File: pkg/core/dao/dao.go [hunk 13]
```diff
@@ -510,16 +510,28 @@ func makeStateRootKey(height uint32) []byte {
...
-		dao.MPT = mpt.NewTrie(nil, dao.Store)
-		return nil
+		dao.MPT = mpt.NewTrie(nil, enableRefCount, dao.Store)
+		var val byte
+		if enableRefCount {
+			val = 1
+		}
+		return dao.Store.Put(gcKey, []byte{val})
+	}
+	var hasGC bool
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/dao/dao.go:523]: Fixed.

File: pkg/core/mpt/node.go [hunk 14]
```diff
@@ -45,6 +51,22 @@ func (n *NodeObject) DecodeBinary(r *io.BinReader) {
 	n.Node = DecodeNodeWithType(r)
 }
 
+func (n NodeObject) getNode() Node {
+	return n.Node
+}
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/mpt/node.go:56]: Removed.

File: pkg/core/mpt/trie.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/mpt/trie.go:21]: Fixed.

File: pkg/core/mpt/trie.go (hunk 10 above)
Comment (fyrchik) [reply pkg/core/mpt/trie.go:377]: Fixed.

Comment (fyrchik) [reply pkg/core/mpt/trie.go:377]: Removing items from map if they were not changed or removed from the store.

File: pkg/core/mpt/node.go [hunk 15]
```diff
@@ -28,6 +27,12 @@ type NodeObject struct {
 	Node
 }
 
+// extendedNodeObject extends NodeObject with refcounting-related info.
+type extendedNodeObject struct {
```
Comment (roman-khimov) [inline pkg/core/mpt/node.go:31]: I don't see it being use anymore.

**Review by roman-khimov — CHANGES_REQUESTED** [review]

File: pkg/core/mpt/trie.go (hunk 10 above)
Comment (roman-khimov) [reply pkg/core/mpt/trie.go]: You need to drop items from this cache during `Collapse`, otherwise you're effectively leaking memory.

File: pkg/core/mpt/trie.go [hunk 16]
```diff
@@ -314,32 +361,91 @@ func makeStorageKey(mptKey []byte) []byte {
...
+	var removed []util.Uint256
+	for h, node := range t.refcount {
+		if node.refcount != 0 {
+			if node.bytes == nil {
+				panic("item not in trie")
+			}
+			if t.refcountEnabled {
+				node.initial = t.updateRefCount(h, node)
+				if node.initial == 0 {
+					removed = append(removed, h)
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/mpt/trie.go:373]: Fixed.

File: pkg/core/mpt/trie.go [hunk 17]
```diff
@@ -314,32 +361,91 @@ func makeStorageKey(mptKey []byte) []byte {
...
+	for i := range removed {
+		delete(t.refcount, removed[i])
+	}
 }
 
-func (t *Trie) flush(node Node) {
-	if node.IsFlushed() {
-		return
+// updateRefCount should be called only when refcounting is enabled.
+func (t *Trie) updateRefCount(h util.Uint256, node *cachedNode) int {
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/mpt/trie.go:389]: It was here mostly because we already have an item when calling this.

File: pkg/core/mpt/node.go (hunk 15 above)
Comment (fyrchik) [reply pkg/core/mpt/node.go:31]: Fixed.

File: pkg/core/mpt/trie.go (hunk 17 above)
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/mpt/trie.go:389]: Fixed.

File: pkg/core/mpt/trie.go [hunk 18]
```diff
@@ -13,7 +15,13 @@ import (
 type Trie struct {
 	Store *storage.MemCachedStore
 
-	root Node
+	root            Node
+	refcountEnabled bool
+	refcount        map[util.Uint256]*cachedNode
```
[the remark this answers is not in the document]
Comment (fyrchik) [reply pkg/core/mpt/trie.go:20]: It's 40 (24 for slice + 16 for ints) and requires copy on every `addRef/removeRef`. I will bench both variants.
Fixed.

**Comment by fyrchik** [comment]
```
# With pointers, refcount off
real    108m18,157s
user    197m42,104s
sys     15m6,246s
12G     chains

# With pointers, refcount on
real    118m18,436s
user    226m17,145s
sys     17m5,012s
8,3G    chains

# No pointers, refcount off
real    114m21,650s
user    213m47,008s
sys     15m48,879s
12G     chains

# No pointers, refcount on
Failed with panic early, will investigate.
```

But when comparing 3 first there is a ~6-7% performance decrease going from implementation using pointers to non-pointers one.

File: pkg/core/mpt/trie.go (hunk 10 above)
Comment (fyrchik) [reply pkg/core/mpt/trie.go]: What do you mean? The map is updated in `Collapse`.

**Comment by roman-khimov** [comment]
OK, it looks like pointers win here despite numerous microallocations.

**Review by roman-khimov — APPROVED** [review]
